buildSrc/src/main/java/io/crate/gradle/plugins/jdk/Jdk.java

```diff
@@ -0,0 +1,176 @@
+    private static final Pattern VERSION_PATTERN = Pattern.compile(
+        "(\\d+)(\\.\\d+\\.\\d+)?\\+(\\d+(?:\\.\\d+)?)(@([a-f0-9]{32}))?");
+
+    private final String name;
+    private final Configuration configuration;
+
+    private final Property<String> vendor;
+    private final Property<String> version;
+    private final Property<String> platform;
+    private String baseVersion;
+    private String major;
+    private String build;
+    private String hash;
+
+    Jdk(String name, Configuration configuration, ObjectFactory objectFactory) {
+        this.name = name;
+        this.configuration = configuration;
+        this.vendor = objectFactory.property(String.class);
+        this.version = objectFactory.property(String.class);
+        this.platform = objectFactory.property(String.class);
+    }
+
+    public String name() {
+        return name;
+    }
+
+    public String vendor() {
+        return vendor.get();
+    }
+
+    public void setVendor(final String vendor) {
+        if (!ALLOWED_VENDORS.contains(vendor)) {
+            throw new IllegalArgumentException(
+                "unknown vendor [" + vendor + "] for jdk [" + name + "], " +
+                "must be one of " + ALLOWED_VENDORS);
+        }
+        this.vendor.set(vendor);
+    }
+
+    public String version() {
+        return version.get();
+    }
+
+    public void setVersion(String version) {
+        var versionMatcher = VERSION_PATTERN.matcher(version);
+        if (!versionMatcher.matches()) {
+            throw new IllegalArgumentException("malformed version [" + version + "] for jdk [" + name + "]");
+        }
+        baseVersion = versionMatcher.group(1) + versionMatcher.group(2);
+        major = versionMatcher.group(1);
+        build = versionMatcher.group(3);
+        hash = versionMatcher.group(5);
+        this.version.set(version);
+    }
+
+    public String platform() {
+        return platform.get();
+    }
+
+    public void setPlatform(String platform) {
+        if (!ALLOWED_PLATFORMS.contains(platform)) {
+            throw new IllegalArgumentException(
+                "unknown platform [" + platform + "] for jdk [" + name + "], " +
+                "must be one of " + ALLOWED_PLATFORMS
+            );
+        }
+        this.platform.set(platform);
+    }
+
+    public String baseVersion() {
+        return baseVersion;
+    }
+
+    public String major() {
+        return major;
+    }
+
+    public String build() {
+        return build;
+    }
+
+    public String hash() {
+        return hash;
+    }
+
+    public String path() {
+        return configuration.getSingleFile().toString();
+    }
+
+    public Configuration configuration() {
+        return configuration;
+    }
+
+    @Override
+    public String toString() {
+        return path();
+    }
+
+    @Override
+    public TaskDependency getBuildDependencies() {
+        return configuration.getBuildDependencies();
+    }
+
+    public Object getBinJavaPath() {
```

Comment:
Is this some gradle specific thing that this can't return a `String` directly?

Reply:
It doesn't work for some gradle reason. Seems like gradle relies on the `toString` call, e.g. https://docs.gradle.org/current/javadoc/org/gradle/api/Project.html#getVersion-- But `String` also has `toString`, i am not sure, have to look a bit why it is the case.